tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc

```diff
@@ -524,18 +524,34 @@ Status TrtShapeOptimizationProfile::SetInputShapeBinding(
       "TrtShapeOptimizationProfile::SetInputShapeBinding",
       tensorflow::profiler::TraceMeLevel::kInfo);
   if (cuda_engine->isShapeBinding(binding_index)) {
-    // Input shape binding data has to be in host memory. That is the reason
-    // we can't use input_tensor.flat().data(). which contains the same
-    // values in device memory. Instead, we use data that was copied to host
-    // by CollectShapeValues.
-    VLOG(2) << "Setting input shape binding for idx " << binding_index
-            << ", with values "
-            << DebugString(actual_shape_values_.at(input_index));
-    bool ret = exec_context->setInputShapeBinding(
-        binding_index, actual_shape_values_.at(input_index).d);
-    if (!ret) {
-      return errors::Internal("Could not set input shape binding for idx ",
-                              binding_index);
+    const auto& newDims = actual_shape_values_.at(input_index);
+    int set_shape = 1;
+    if (exec_context->getOptimizationProfile() != -1 &&
+        (cuda_engine->bindingIsInput(binding_index) ||
+        exec_context->allInputDimensionsSpecified() &&
+        exec_context->allInputShapesSpecified())) {
+      const auto oldDims = exec_context->getBindingDimensions(binding_index);
+      if (oldDims.nbDims == newDims.nbDims) {
```

Comment:
You should compare not only the number of dimensions (nbDims), but also the actual dimensions (dims.d). You can simply use the following, because we have a comparison operator implemented in common/utils.h:
```suggestion
      if (oldDims == newDims) {
```

Reply:
My suggestion above is incorrect. Since `newDims` is from `actual_shape_values_` ([defined here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc#L172-L174)), there nbDims just means the total number of elements. Therefore you would need something like 
```suggestion
      DimsAdapter da(oldDims);
      if (da.Volume() == newDims.nbDims) {
```